### DISCUSSION TITLE
How to throw an exception that returns 302 redirect http response

### DISCUSSION THREAD

**Discussion by yevon** [PostgREST/postgrest#2486] [Q&A] [answered]
I'm trying to implement a pre-request procedure that performs an exception with a 302, how can this be achieved? I don't see a mapped postgres error for 302 responses in the docs:

<html>
<body>

08* | 503 | pg connection err
09* | 500 | triggered action exception
0L* | 403 | invalid grantor
0P* | 403 | invalid role specification
23503 | 409 | foreign key violation
23505 | 409 | uniqueness violation
25006 | 405 | read only sql transaction
25* | 500 | invalid transaction state
28* | 403 | invalid auth specification
2D* | 500 | invalid transaction termination
38* | 500 | external routine exception
39* | 500 | external routine invocation
3B* | 500 | savepoint exception
40* | 500 | transaction rollback
53* | 503 | insufficient resources
54* | 413 | too complex
55* | 500 | obj not in prerequisite state
57* | 500 | operator intervention
58* | 500 | system error
F0* | 500 | config file error
HV* | 500 | foreign data wrapper error
P0001 | 400 | default code for “raise”
P0* | 500 | PL/pgSQL error
XX* | 500 | internal error
42883 | 404 | undefined function
42P01 | 404 | undefined table
42501 | if authenticated 403,else 401 | insufficient privileges
other | 400

<br class="Apple-interchange-newline">
</body>
</html>

**Comment by yevon** [comment]
I tried to modify the response status before calling the exception, but it doesn't persist.


      perform set_config('response.status', '302', true);
      RAISE insufficient_privilege USING
          message = 'Request validation failed',
          detail = 'Session expired, please log in.',
          hint = 'UNAUTHORIZED';

**Comment by yevon** [comment, marked as answer]
I found the way for returning the PT302 code, but if I raise an exception I can't set reponse headers as they are rolled back:

This returns 302 http code, but ignores the location response header:

```
       l_redirect_location := '[{"Location": "/rpc/token_refresh?redirect_url=%redirect_url"}]';
       l_redirect_location := replace(l_redirect_location, '%redirect_url', urlencode(app.http_get_header(p_header_name => 'original-url')));
        perform set_config('response.headers', l_redirect_location, true);     

        RAISE sqlstate 'PT302' using
        message = 'Found',
        detail = 'Token expired, please refresh token or log in.',
        hint = 'UNAUTHORIZED';
```

**Reply by steve-chavez** [reply]
Yes the whole transaction will be rolled back, this will reset the local GUCs.

Hmm, the only way I see we could make this work is by (ab)using the EXCEPTION message? Like

```sql
RAISE EXCEPTION 'headers: %, status: %', '{"X-Header": "str"}', '302' USING HINT = 'a hint...';
```
@yevon WDYT?

---

It does work:
```sql
DO $$BEGIN RAISE EXCEPTION 'headers: %, status: %', '{"X-Header": "str"}', '302' USING HINT = 'a hint..'; END$$;

ERROR:  headers: {"X-Header": "str"}, status: 302
HINT:  a hint..
CONTEXT:  PL/pgSQL function inline_code_block line 1 at RAISE
```

**Reply by yevon** [reply]
I will try this out thanks! It looks quite hacky. I was thinking that It would be nice being able to define a pre-request that returns a boolean, that would tell if execution must proceed or not. This way pre-request could be used for cases like this, without needing to raise exceptions if you want to persist some changes.

**Reply by wolfgangwalther** [reply]
We already have the custom `PTxxx` error codes that map to a response status code. How about we add a feature to support a custom `PGRST` SQLSTATE error, where both `message` and `detail` take JSON. `message` would contain the response body for this error (i.e. message, detail, hint etc. all in there) and `detail` would contain a list of headers to set?

**Reply by yevon** [reply]
Yes it would be nice indeed, at the end the objective is being able to fully control the response in case of an exception: http code, body response, response headers.

**Comment by yevon** [comment]
Hi @steve-chavez, I thought that you meant that this was working already but you just proposed to implement this change? I tried it but it is just returning a 400 error and the message. 

Having a pre-request returning a boolean would be nicer I think, as we could use already existent functions to set headers and state. Or maybe just defining an optional new a pre-request-validation function that would return a boolean, specifying if the request should proceed or not.

It would work like this:

1. Execute pre-request-validation: Returns a boolean, true if it should continue, false if it should abort. (You could raise an exception if you want too). If you return a false without an http status code set, it would just return a default 400 "pre request validation failed".
2. Execute current pre-request procedure (For backwards compatibility, or maybe we don't need two separate functions for this)
3. Continue execution

**Reply by wolfgangwalther** [reply]
Once we have #2295 we could indeed get a return from the pre-request function. I wouldn't take a boolean, though, because that would require us to run another query to get those headers. How about the headers etc. directly from the pre-request function?

However, that would require to run the main query first, and then rollback if anything comes back from the pre-request later on. We might be better of with extending how to throw errors as @steve-chavez suggested.

**Reply by yevon** [reply]
Nice! If we can fully control the response through exceptions it would do the trick! Should I create an issue for this?

**Reply by steve-chavez** [reply]
@yevon Sure, go for it. I like the interface Wolfgang mentioned on https://github.com/PostgREST/postgrest/discussions/2486#discussioncomment-3742773 since we already use `PGRST` for error codes.

**Comment by yevon** [comment]
I'm in the first steps implementing this, but I'm having some trouble when I was trying to test this, when trying to start a local postgrest environment with the nix-shell. 

I installed a local database, executed the nix-shell and run a "postgrest-run" command, I can't figure out what parameters does it use to connect, what database, anonymous user etc. Where I cant edit those params? Thanks! Is there any documentation about how to do this?

I figured out that it was trying to connect to a local database with my username, but not much more than that.

**Reply by wolfgangwalther** [reply]
You don't need to install a local database. When using the nix environment, a dev database is set up automatically when you run the various `postgrest-xxx` commands. You can also run any command, i.e. psql, with a temporary dev database, by doing `postgrest-with-postgresql-15 psql` etc.

**Comment by yevon** [comment]
Solved! I just found a readme file. I just had to create a .conf file with proper parameters and just call it with "postgrest-run db.conf"
